### DISCUSSION TITLE
How to show local images and/or videos?

### DISCUSSION THREAD

**Discussion by Hayoung93** [zauberzeug/nicegui#907] [Q&A] [answered]
### Question

Hello, thank you for the great work. I faced following problem while using NiceGUI. Can I get some advise? Thanks.

I want to demonstrate some local images and/or videos using NiceGUI (possibly using `ui.upload` together).

I refer to another discussion and added `app.add_static_files`, but I'm keep failing.

Here's my code:

``` python
from nicegui import ui, app
app.add_static_files("/workspace/static", ".")
v = ui.video("/workspace/static/xx.mp4")
v.on('ended', lambda _: ui.notify('Video playback completed'))
ui.image("/workspace/static/xx.png")
ui.run(host="0.0.0.0", port=8080)
```
And the error message: 
```
http://localhost:18080/workspace/static/xx.png not found
http://localhost:18080/workspace/static/xx.mp4 not found
```

The UI keeps adding local address before the actual file path.

Also, I'd tried to upload local files using `ui.upload`:
```python
ui.upload(max_file_size=100000000, max_total_size=100000000, max_files=10, auto_upload=True,
          on_upload=lambda e: ui.notify(f'Uploaded {e.name}')).classes('max-w-full')
```
However, I have no idea how to call uploaded content since the url kept changes every time, for example, `/_nicegui/client/xxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxx/upload/x`.

When I fixed the uploading url to given argument by modifying `nicegui/elements/upload.py`, and execute this code:
``` python
ui.upload(max_file_size=100000000, max_total_size=100000000, max_files=10, auto_upload=True, url="0.0.0.0:8080/data"
          on_upload=lambda e: ui.notify(f'Uploaded {e.name}')).classes('max-w-full')
```
gives `http://localhost:18080/0.0.0.0:8080/data not found`.

-------------------
The code was executed inside of a docker container, which is created via `docker run -it --name nicegui -p 18080:8080 -p 18081:8081 pytorch/pytorch:1.12.1-cuda11.3-cudnn8-devel` command.

**Comment by rodja** [comment, marked as answer]
Your code with video playback works for me as long as the `xx.mp4` and `xx.png` are located in the directory from where I start the NiceGUI app. Keep in mind that `ui.add_static_files` takes the url path in the first argument and the local directory which holds the files you want to make available in the second parameter.

To save the uploaded content to disc you should do something similar as https://nicegui.io/documentation/upload#show_file_content. The data are available in `e.content.read()`. The changing urls are only used internally to ensure that multiple uploads can be performed in parallel.

**Reply by Hayoung93** [reply]
Thank you for the answer. I should have check the parameters of `add_static_files` :smiling_face_with_tear:

**Reply by falkoschindler** [reply]
Note that it's actually `app.add_static_files` rather than `ui.add_static_files`.

**Comment by rodja** [comment]
We just released [v1.2.20](https://github.com/zauberzeug/nicegui/releases/tag/v1.2.20) which introduces `app.add_media_files` for streaming support and also allows you to pass a local file directly to `ui.video`, `ui.audio`, `ui.image` and `ui.interactive_image`. It will automatically be served correctly.
